clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -2286,6 +2286,9 @@ ExprResult Sema::BuildCXXNew(SourceRange Range, bool UseGlobal,
   bool PassAlignment = getLangOpts().AlignedAllocation &&
                        Alignment > NewAlignment;
 
+  if (CheckArgsForPlaceholders(PlacementArgs))
```

Comment:
This likely needs a good collection of tests to show what it changes for NON-property related 'new' operators.

Reply:
I’ll take a look at what other expressions are actually affected by this other than `MSPropertyRefExpr`s. (or do you mean we should just add more placement-new tests in general? Because I would assume we already have a bunch of those, but I haven’t checked to be fair?)

Reply:
A bit of both, this should check the behavior of other expression types.  This changes general 'placement new' behaviors, so it looks like it changes existing placement new behaviors that are seemingly untested, so should be showing off  THOSE as well.

Reply:
Makes sense. It seems like this affects only a small number of builtin types, so in theory, adding tests for every expression that could possibly have those types ought to do (if we don’t already have tests for those); I’m currently collecting a list of everything that could be affected by this.

Reply:
The only expressions that this seems to affect is expressions for which `isPlaceholderToRemoveAsArg` returns `true` by passing them to `CheckPlaceholderExpr`; this only affects a handful of types:

- `BuiltinType::PseudoObject`: MSProperty (subscript) properties and ObjC (subscript) properties; the latter are only relevant for Objective-C++ in this case. I’ll add tests for the latter as well (to the best of my abilities as I have never written a line of Objective-C or Objective-C++ before in my life). The comment in `BuiltinTypes.def` also mentions ‘VS.NET's `__property` declarations’, but I couldn’t find anything about `__property` in the rest of the codebase, and I’m candidly not entirely sure what this is in reference to, so I’m only adding tests for Objective-C++ for now.
    
- `BuiltinType::UnknownAny`: For this, `CheckPlaceholderExpr` just emits an error, and I personally don’t see a way in which this would ever be valid as a placement-arg to a new expression, but I could be wrong here. Should we add tests for this case?

- `BuiltinType::BoundMember`: `s.foo` and friends, where `foo` is a non-static member function; these are already invalid in placement-args, and `CheckPlaceholderExpr` similarly issues what looks to be the exact same error; we don’t seem to have any tests that check for this error, however, so I’ll add some for this as well.

- `BuiltinType::BuiltinFn`: Only `DeclRefExpr`s are affected here as this simply issues an error otherwise (which should be fine because passing a builtin function as a function argument doesn’t make sense anyway); if it is a builtin such as `__builtin_memset`, then this pr also ends up improving the error message from ‘no known conversion from '<builtin fn type>'’ to ‘builtin functions must be directly called’.

    If it is a `DeclRefExpr`, there are two cases to consider: the first is the MS `__noop` builtin, which w/o this pr simply errors, but now, it is implicitly converted to a `CallExpr` that returns an `int` (which emits 0 in CG); this matches MSVCs behaviour, but I doubt we have a test for that, so I’ll add one as well.

    The only other case that is affected here is if we have a standard-namespace builtin (e.g. `std::move`, `std::forward_like`, which before C++20 resulted in a warning, since C++ in an error; this behaviour is unaffected by this change, but I’ll some tests for this as well because from what I can tell, there currently are no tests for this diagnostic at all.


- `BuiltinType::IncompleteMatrixIdx`: From what I can tell, this is used for [matrix types](https://clang.llvm.org/docs/MatrixTypes.html), specifically, for the type of an expression such as `a[1]` where `a` is a matrix type because matrix types always require two subscripts; since an expression of this type as a placement-arg is thus always invalid, this should error, but previously, we were emitting a less-than-ideal error (e.g. ‘no known conversion from '<incomplete matrix index type>' to 'int'’), whereas with this pr, we now get the error ‘single subscript expressions are not allowed for matrix values’. There does not seem to be a test for this that involves placement-new, so I’ll add a few as well.

- `BuiltinType::OMPArraySection`, `BuiltinType::OMPArrayShaping`, and `BuiltinType::OMPIterator`: To my understanding, these cannot occur outside of OMP `#pragma`s, so they’re not relevant here.

Reply:
tl;dr Overall, it seems like this pr also improves some error messages as well as support for MSVC’s `__noop` builtin. I’ll add some tests for the cases above (though again, for some of them, (e.g. the UnknownAny case), I’m not sure whether that’s necessary, so please let me know whether I should add some for those as well).

Reply:
Oh, and Objective-C++ properties were suffering from the same problem, so this now also fixes those in new-expressions.

Reply:
Alright, just got done adding some tests. Just let me know if there are more things that we should consider testing. 

Also, regarding Objective-C++ properties: I tried following the examples in `clang/test/AST/ast-dump-expr-json.m`, but it didn’t seem to want to treat the subscripts as anything other than pointer arithmetic. It also doesn’t help that I’m not familiar w/ Objective-C/Objective-C++ at all. Could it be that Objective-C subscript properties are disabled in Objective-C++ mode?